loadbalancer/algorithm.go

```diff
@@ -168,32 +163,23 @@ func (r *roundRobin) Apply(ctx *routing.LBContext) routing.LBEndpoint {
 		return ctx.Route.LBEndpoints[0]
 	}
 
-	r.mx.Lock()
-	defer r.mx.Unlock()
-	r.index = (r.index + 1) % len(ctx.Route.LBEndpoints)
+	index := int(atomic.AddInt64(&r.index, 1) % int64(len(ctx.Route.LBEndpoints)))
 
 	if ctx.Route.LBFadeInDuration <= 0 {
-		return ctx.Route.LBEndpoints[r.index]
+		return ctx.Route.LBEndpoints[index]
 	}
 
-	return withFadeIn(r.rnd, ctx, r.notFadingIndexes, r.index, r)
```

Comment:
I think we misunderstood the idea of this `r.notFadingIndexes`: yes here we pass a slice by value and its length is always zero but capacity is equal to `len(endpoints)` (see constructor). 
`withFadeIn` writes to its underlying array and then passes to `shiftToRemaining` which uses values from underlying array.
So the idea of `r.notFadingIndexes` is to act like a scratch buffer to avoid allocations on each request (https://github.com/zalando/skipper/pull/2412/files#r1238234491).
The real problem is that access to this scratch buffer is not synchronized.

Reply:
The same is true for `a.fadingWeights` - it is also used as a scratch buffer.

Reply:
```
access to this scratch buffer is not synchronized.
```
I still don't understand why should we synchronize access to the struct field we never write to.

Reply:
Another thought is that it's possible to synchronize access by per-loadbalancer lock like this, but this way performance will suffer while concurrent access, much more that with suboptimal memory (re)allocation.

```
+       ch.mx.Lock()
+       defer ch.mx.Unlock()
        return withFadeIn(ch.rand, ctx, ch.notFadingIndexes, choice, ch)
 }
```

Reply:
```
 is to act like a scratch buffer to avoid allocations on each request
```
One more thought - this idea sounds legit in theory, but I haven't seen any performance advance by using such scratch buffers.

Reply:
> One more thought - this idea sounds legit in theory, but in practice I haven't seen any performance advance by using such scratch buffers.

I think you saw https://github.com/zalando/skipper/pull/2412/files#r1238267164